mrmustard/physics/fock.py

```diff
@@ -467,3 +469,204 @@ def trace(dm, keep: List[int]):
+            with number of photons ``i`` evaluated at position ``q[j]``, i.e., `\psi_i(q_j)`.
+
+    .. details::
+
+        .. admonition:: Definition
+            :class: defn
+
+        The q-quadrature eigenstates are defined as
+
+        .. math::
+
+            \psi_n(x) = 1/sqrt[2^n n!](\frac{\omega}{\pi \hbar})^{1/4}
+                \exp{-\frac{\omega}{2\hbar} x^2} H_n(\sqrt{\frac{\omega}{\pi}} x)
+
+        where :math:`H_n(x)` is the (physicists) `n`-th Hermite polynomial.
+    """
+    omega_over_hbar = math.cast(1 / settings.HBAR, "float64")
+    x_tensor = math.sqrt(omega_over_hbar) * math.cast(q, "float64")  # unit-less vector
+
+    # prefactor term (\Omega/\hbar \pi)**(1/4) * 1 / sqrt(2**n)
+    prefactor = (omega_over_hbar / np.pi) ** (1 / 4) * math.sqrt(2 ** (-math.arange(0, cutoff)))
+
+    # Renormalized physicist hermite polys: Hn / sqrt(n!)
+    R = math.astensor(2 * np.ones([1, 1]))  # to get the physicist polys
+
+    def f_hermite_polys(xi):
+        return math.hermite_renormalized(R, 2 * math.astensor([xi]), 1, cutoff)
+
+    hermite_polys = math.cast(math.map_fn(f_hermite_polys, x_tensor), "float64")
+
+    # wavefunction
+    psi = math.exp(-(x_tensor**2 / 2)) * math.transpose(prefactor * hermite_polys)
+    return psi
+
+
+@lru_cache
+def estimate_dx(cutoff, period_resolution=20):
+    r"""Estimates a suitable quadrature discretization interval `dx`. Uses the fact
+    that Fock state `n` oscillates with angular frequency :math:`\sqrt{2(n + 1)}`,
+    which follows from the relation
+
+    .. math::
+
+            \psi^{[n]}'(q) = q - sqrt(2*(n + 1))*\psi^{[n+1]}(q)
+
+    by setting q = 0, and approximating the oscillation amplitude by `\psi^{[n+1]}(0)
+
+    Ref: https://en.wikipedia.org/wiki/Hermite_polynomials#Hermite_functions
+
+    Args
+        cutoff (int): Fock cutoff
+        period_resolution (int): Number of points used to sample one Fock
+            wavefunction oscillation. Larger values yields better approximations
+            and thus smaller `dx`.
+
+    Returns
+        (float): discretization value of quadrature
+    """
+    fock_cutoff_frequency = np.sqrt(2 * (cutoff + 1))
+    fock_cutoff_period = 2 * np.pi / fock_cutoff_frequency
+    dx_estimate = fock_cutoff_period / period_resolution
+    return dx_estimate
+
+
+@lru_cache
+def estimate_xmax(cutoff, minimum=5):
+    r"""Estimates a suitable quadrature axis length
+
+    Args
+        cutoff (int): Fock cutoff
+        minimum (float): Minimum value of the returned xmax
+
+    Returns
+        (float): maximum quadrature value
+    """
+    if cutoff == 0:
+        xmax_estimate = 3
+    else:
+        # maximum q for a classical particle with energy n=cutoff
+        classical_endpoint = np.sqrt(2 * cutoff)
+        # approximate probability of finding particle outside classical region
+        excess_probability = 1 / (7.464 * cutoff ** (1 / 3))
```

Comment:
what is the 7.464 constant? 🤔

Reply:
The probability of finding the particle outside the classically allowed region is approximately

![image](https://user-images.githubusercontent.com/675763/202514993-4e745a3d-e6ca-40a2-978e-33de178b841e.png)

i.e., one gets there by taking the probability density integrated between the classical boundaries and use the asymptotics of the Hermite polys [[wikipedia](https://en.wikipedia.org/wiki/Quantum_harmonic_oscillator#Highly_excited_states)].